File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/mouse/LeftClickGesture.java
package net.kdt.pojavlaunch.customcontrols.mouse;

import static org.lwjgl.glfw.CallbackBridge.sendMouseButton;

import android.os.Handler;

import com.arata.yukarilauncher.setting.AllStaticSettings;

import net.kdt.pojavlaunch.LwjglGlfwKeycode;
import net.kdt.pojavlaunch.Tools;
import net.kdt.pojavlaunch.utils.MathUtils;

import org.lwjgl.glfw.CallbackBridge;

public class LeftClickGesture extends ValidatorGesture {
    public static final int FINGER_STILL_THRESHOLD = (int) Tools.dpToPx(9);
    private float mGestureStartX, mGestureStartY, mGestureEndX, mGestureEndY;
    private boolean mMouseActivated;
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public LeftClickGesture(Handler handler) {
        super(handler);
    }
/**
 * 「input Event」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    public final void inputEvent() {
        if(submit()) {
            mGestureStartX = mGestureEndX = CallbackBridge.mouseX;
            mGestureStartY = mGestureEndY = CallbackBridge.mouseY;
        }
    }
/**
 * 「GestureDelay」の値を取得します。
 */
    @Override
    protected int getGestureDelay() {
        return AllStaticSettings.timeLongPressTrigger;
    }
/**
 * 「check And Trigger」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    @Override
    public boolean checkAndTrigger() {
        boolean fingerStill = LeftClickGesture.isFingerStill(mGestureStartX, mGestureStartY, mGestureEndX, mGestureEndY, FINGER_STILL_THRESHOLD);
        // If the finger is still, fire the gesture.
        if(fingerStill) {
            sendMouseButton(LwjglGlfwKeycode.GLFW_MOUSE_BUTTON_LEFT, true);
            mMouseActivated = true;
        }
        // Otherwise, don't click but still keep it active
        return true;
    }
/**
 * 「on Gesture Cancelled」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    @Override
    public void onGestureCancelled(boolean isSwitching) {
        if(mMouseActivated) {
            sendMouseButton(LwjglGlfwKeycode.GLFW_MOUSE_BUTTON_LEFT, false);
            mMouseActivated = false;
        }
    }
/**
 * 「Motion」の値を設定します。
 */
    public void setMotion(float deltaX, float deltaY) {
        mGestureEndX += deltaX;
        mGestureEndY += deltaY;
    }

    /**
     * CallbackBridgeのmouseX/mouseYと比較して指が静止しているか確認します。
     * @param startX ジェスチャーの開始X座標
     * @param startY ジェスチャーの開始Y座標
     * @return 指の位置が「静止」とみなされるかどうか
     */
    public static boolean isFingerStill(float startX, float startY, float threshold) {
        return MathUtils.dist(
                CallbackBridge.mouseX,
                CallbackBridge.mouseY,
                startX,
                startY
        ) <= threshold;
    }
/**
 * このオブジェクトが「FingerStill」状態であるかを判定します。
 */
    public static boolean isFingerStill(float startX, float startY, float endX, float endY, float threshold) {
        return MathUtils.dist(
                endX,
                endY,
                startX,
                startY
        ) <= threshold;
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/mouse/LeftClickGesture.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols.mouse

import android.os.Handler
import com.arata.yukarilauncher.setting.AllStaticSettings
import com.arata.yukarilauncher.utils.LwjglGlfwKeycode
import com.arata.yukarilauncher.Tools
import com.arata.yukarilauncher.utils.MathUtils
import org.lwjgl.glfw.CallbackBridge

/** 左クリックの長押しジェスチャーを検出・処理するクラス。 */
class LeftClickGesture(handler: Handler) : ValidatorGesture(handler) {
    companion object {
        /** 指が静止していると判定する閾値（px）。 */
        val FINGER_STILL_THRESHOLD = Tools.dpToPx(9f).toInt()

        /**
         * 指が静止しているか判定します（現在位置と開始位置を比較）。
         * @param startX 開始X座標
         * @param startY 開始Y座標
         * @param threshold 静止判定の閾値
         * @return 静止している場合はtrue
         */
        fun isFingerStill(startX: Float, startY: Float, threshold: Float): Boolean {
            return MathUtils.dist(
                CallbackBridge.mouseX,
                CallbackBridge.mouseY,
                startX,
                startY
            ) <= threshold
        }

        /**
         * 指が静止しているか判定します（2点間の距離）。
         * @param startX 開始X座標
         * @param startY 開始Y座標
         * @param endX 終了X座標
         * @param endY 終了Y座標
         * @param threshold 静止判定の閾値
         * @return 静止している場合はtrue
         */
        fun isFingerStill(startX: Float, startY: Float, endX: Float, endY: Float, threshold: Float): Boolean {
            return MathUtils.dist(
                endX,
                endY,
                startX,
                startY
            ) <= threshold
        }
    }

    private var mGestureStartX = 0f
    private var mGestureStartY = 0f
    private var mGestureEndX = 0f
    private var mGestureEndY = 0f
    private var mMouseActivated = false

    /** 左クリックの入力イベントを送信します。 */
    fun inputEvent() {
        if (submit()) {
            mGestureStartX = CallbackBridge.mouseX
            mGestureStartY = CallbackBridge.mouseY
            mGestureEndX = CallbackBridge.mouseX
            mGestureEndY = CallbackBridge.mouseY
        }
    }

    override fun getGestureDelay(): Int = AllStaticSettings.timeLongPressTrigger

    override fun checkAndTrigger(): Boolean {
        val fingerStill = isFingerStill(mGestureStartX, mGestureStartY, mGestureEndX, mGestureEndY, FINGER_STILL_THRESHOLD.toFloat())
        if (fingerStill) {
            CallbackBridge.sendMouseButton(LwjglGlfwKeycode.GLFW_MOUSE_BUTTON_LEFT.toInt(), true)
            mMouseActivated = true
        }
        return true
    }

    override fun onGestureCancelled(isSwitching: Boolean) {
        if (mMouseActivated) {
            CallbackBridge.sendMouseButton(LwjglGlfwKeycode.GLFW_MOUSE_BUTTON_LEFT.toInt(), false)
            mMouseActivated = false
        }
    }

    /**
     * マウスの移動量を設定します。
     * @param deltaX X方向の移動量
     * @param deltaY Y方向の移動量
     */
    fun setMotion(deltaX: Float, deltaY: Float) {
        mGestureEndX += deltaX
        mGestureEndY += deltaY
    }
}